Sources/PackageRegistryTool/PackageRegistryTool+ArchiveSign.swift

```diff
@@ -0,0 +1,469 @@
+        }
+
+        var signedMetadata: SignedItem? = .none
+        if let metadataPath {
+            observabilityScope.emit(info: "signing metadata at '\(metadataPath)'")
+            let metadata = try localFileSystem.readFileContents(metadataPath).contents
+            let metadataSignature = try SignatureProvider.sign(
+                content: metadata,
+                identity: signingIdentity,
+                intermediateCertificates: intermediateCertificates,
+                format: signatureFormat,
+                observabilityScope: observabilityScope
+            )
+            let metadataSignaturePath = workingDirectory.appending("\(packageIdentity)-\(packageVersion)-metadata.sig")
+            try fileSystem.writeFileContents(metadataSignaturePath) { stream in
+                stream.write(metadataSignature)
+            }
+            signedMetadata = .init(path: metadataPath, signature: metadataSignature)
+        }
+
+        return ArchiveAndSignResult(
+            archive: .init(path: archivePath, signature: archiveSignature),
+            signedManifests: manifests,
+            metadata: signedMetadata
+        )
+    }
+
+    private static func signingIdentityAndIntermediateCertificates(
+        mode: SigningMode,
+        signatureFormat: SignatureFormat,
+        fileSystem: FileSystem,
+        observabilityScope: ObservabilityScope
+    ) throws -> (SigningIdentity, [[UInt8]]) {
+        let signingIdentity: SigningIdentity
+        let intermediateCertificates: [[UInt8]]
+        switch mode {
+        case .identityStore(let label, let intermediateCertPaths):
+            let signingIdentityStore = SigningIdentityStore(observabilityScope: observabilityScope)
+            let matches = signingIdentityStore.find(by: label)
+            guard let identity = matches.first else {
+                throw StringError("'\(label)' not found in the system identity store.")
+            }
+            // TODO: let user choose if there is more than one match?
+            signingIdentity = identity
+            intermediateCertificates = try intermediateCertPaths.map { try fileSystem.readFileContents($0).contents }
+        case .certificate(let certPath, let intermediateCertPaths, let privateKeyPath):
+            let certificate = try fileSystem.readFileContents(certPath).contents
+            let privateKey = try fileSystem.readFileContents(privateKeyPath).contents
+            signingIdentity = try SwiftSigningIdentity(
+                derEncodedCertificate: certificate,
+                derEncodedPrivateKey: privateKey,
+                privateKeyType: signatureFormat.signingKeyType
+            )
+            intermediateCertificates = try intermediateCertPaths.map { try fileSystem.readFileContents($0).contents }
+        }
+        return (signingIdentity, intermediateCertificates)
+    }
+
+    private static func findManifests(packageDirectory: AbsolutePath) throws -> [String] {
```

Comment:
is this reusable code we should share?

Reply:
I thought something like this should already exist but I couldn't find it.